### PR TITLE
Add reqwest download backend, remove hyper backends

### LINKED ISSUE #1181: Teach hyper/rustls HTTP backends to resume partial downloads
The curl backend can already do it, but not the two hyper backends. It should just be a matter of threading the parameter through and setting the HTTP [Range header](http://stackoverflow.com/questions/3303029/ddg#18745164).

Test by compiling with `--features=rustls` and `RUSTUP_USE_HYPER` / `RUSTUP_USE_RUSTLS`, and `--verbose`.

No way to actually write test cases for this though.

### LINKED ISSUE #1183: Add a reqwest backend to rustup
rustup currently supports a robust collection of HTTP backends: curl, hyper + native-tls and hyper + rustls. Add a fourth based on reqwest, making sure it has proxy and resume support. This will probably become the default backend.

It should be a matter of following the example in src/download, and adding `RUSTUP_USE_REQWEST` to src/utils/src/utils.rs.

cc https://github.com/rust-lang-nursery/rustup.rs/issues/1181

### PR DESCRIPTION
**Pull request by mattico**
Fixes #1183 
Closes #1181 (hyper backend will no longer exist)

About the lockfile changes:
I had to update native_tls from 1.0.0 to 1.0.4 for reqwest.
`cargo update -p native-tls` complained about not having a valid version of security-framework-sys.
`cargo update -p security-framework-sys` complained about not having a valid version of native-tls.
So I updated everything 🤷‍♂️  .

### CODE REVIEW

**Comment by mattico** [comment]
Ah, doesn't work with file URLs. One moment.

**Comment by mattico** [comment]
Fallout from updates:
MIPS/PPC: backtrace build failed (need to install powerpc-unknown-linux-binutils ?)
Android: net2 build failed (https://github.com/rust-lang-nursery/net2-rs/issues/63)

**Comment by mattico** [comment]
https://github.com/rust-lang-nursery/net2-rs/pull/64
x86_64-linux-android now fails with the same error as the rest. I think the issue is with the docker image (https://github.com/rust-lang/rust-buildbot/tree/master/slaves/linux-cross). The configure script is picking up unprefixed objcopy rather than using the cross one.

**Comment by mattico** [comment]
Rebased.

**Comment by mattico** [comment]
Found a less-hacky way to fix the backtrace-sys build: just use env variables. Now to wait for travis.

**Comment by mattico** [comment]
Hoping the rest of the build works\* , but we're going to hit https://github.com/sfackler/rust-security-framework/issues/43 now.

\* confused why it's not running, the opensource linux backlog is zero right now.

**Comment by mattico** [comment]
Same issue with mio as with net2.

**Comment by mattico** [comment]
I think that was a spurious failure.

**Comment by alexcrichton** [comment]
Thanks for this, and sorry for the churn! If you'd like to rebase one last time I think we should be ready to go in terms of review and looking this over.

**Comment by mattico** [comment]
No worries, @alexcrichton!

Going to try without the objcopy workaround first, since the build image changes may have fixed that issue.

File: src/download/src/lib.rs [hunk 1]
```diff
@@ -690,30 +364,17 @@ pub mod curl {
...
-                    _callback: &Fn(Event) -> Result<()> )
-                    -> Result<()> {
-        Err(ErrorKind::BackendUnavailable("hyper").into())
-    }
-}
-
-#[cfg(not(feature = "rustls-backend"))]
-pub mod rustls {
+#[cfg(not(feature = "reqwest-backend"))]
+pub mod reqwest {
```
Comment (alexcrichton) [inline src/download/src/lib.rs]: Should this be `reqwest_be` or should the above be `reqwest`?

File: src/rustup-utils/src/utils.rs [hunk 2]
```diff
@@ -221,12 +221,9 @@ fn download_file_(url: &Url,
     };
 
     // Download the file
-    let use_hyper_backend = env::var_os("RUSTUP_USE_HYPER").is_some();
```
Comment (alexcrichton) [inline src/rustup-utils/src/utils.rs]: Mind leaving this around for a bit as an alias for using reqwest?

**Comment by alexcrichton** [comment]
@bors: r+

Looks great, thanks!

**Comment by mattico** [comment]
`Cargo.lock` is fun... rebased.

**Pull request merged**